Compile/src/Producers/ValidZllJetsProducer.cc

```diff
@@ -28,32 +81,52 @@ bool ValidZllJetsProducer::DoesJetPass(const KBasicJet* jet, ZJetEvent const& ev
     // check that PUJetID is above configured minimal value
     const KJet* kJet = dynamic_cast<const KJet*>(jet);  // need a KJet for PUJetID, not just a KBasicJet...
     if (kJet) {
-        if (PUJetID == "none") {
-            if (kJet->getTag(PUJetIDModuleName+"fullDiscriminant", event.m_jetMetadata) < minPUJetID) {
+        const double puJetIDValue = kJet->getTag(m_puJetIDMetadataTag, event.m_jetMetadata);
+
+        if (m_puJetIDWorkingPoint == ValidZllJetsProducer::PUJetIDWorkingPoint::FILE) {
+            // lookup bin indices in PUJetID histogram corresponding to the jet pT/absEta
+            const int histBinX = m_puJetIDMinValueHistogram->GetXaxis()->FindFixBin(kJet->p4.Pt());
+            const int histBinY = m_puJetIDMinValueHistogram->GetYaxis()->FindFixBin(std::abs(kJet->p4.Eta()));
+
+            // check for over-/underflows
+            if ((histBinX > 0) && (histBinX <= m_puJetIDMinValueHistogram->GetNbinsX()) &&
+                (histBinY > 0) && (histBinY <= m_puJetIDMinValueHistogram->GetNbinsY())) {
+
+                // if bin is found, check puJetID and veto jet if it fails
+                if (puJetIDValue < m_puJetIDMinValueHistogram->GetBinContent(histBinX, histBinY)) {
+                    return false;
+                }
+            }
+        }
+        else if (m_puJetIDWorkingPoint == ValidZllJetsProducer::PUJetIDWorkingPoint::VALUE) {
+            if (puJetIDValue < settings.GetMinPUJetID()) {
                 return false;
             }
         }
-        else if (PUJetID == "loose") {
-            if (!bool(int(kJet->getTag(PUJetIDModuleName+"fullId", event.m_jetMetadata)) & (1 << 2))) {
+        else if (m_puJetIDWorkingPoint == ValidZllJetsProducer::PUJetIDWorkingPoint::LOOSE) {
```

Comment:
Isn't this again a value cut, as it was in the past? Since there are now eta- and pT-dependent loose, medium and tight working points this is confusing. This should be marked as deprecated, if I understand this correctly.

Reply:
Not exactly. None of the options are deprecated, they just access different things in the inputs and some of them just provide more flexibility. More details below.

The pileup jet ID information for each jet in the MiniAOD comes in two flavors: (1) the full MVA discriminant called `pileupJetID:fullDiscriminant`, and (2) the pre-evaluated pileupJetID working points (tight, medium, loose), all three of which are encoded together as a single `int` called `pileupJetID:fullId`. Basically, when producing the MiniAOD, the (1) is compared to the working points the pass/fail decisions are stored in (2).

In Excalibur, you can either directly check the pre-computed values (2), for which you can set the config `PUJetID` to `tight`, `medium` or `loose`; or you can take the full discriminant value (1) and compare yourself to fixed or pt/eta-dependent working point values, for which you have to set `PUJetID` to `value` or `file`.

Reply:
Okay I understand. Then, it makes perfectly sense to keep it like this.